src/test/shell/bazel/toolchain_test.sh

```diff
@@ -914,4 +914,88 @@ EOF
     expect_log "Selected execution platform //platforms:platform2_4"
 }
 
+function test_default_constraint_values {
+  # Add test constraints and platforms.
+  mkdir -p platforms
+  cat >> platforms/BUILD <<EOF
+package(default_visibility = ['//visibility:public'])
+constraint_setting(name = 'setting1', default_constraint_value = ':value_foo')
+constraint_value(name = 'value_foo', constraint_setting = ':setting1')
+constraint_value(name = 'value_bar', constraint_setting = ':setting1')
+
+platform(
+    name = 'platform_default',
+    constraint_values = [])
+platform(
+    name = 'platform_no_default',
+    constraint_values = [':value_bar'])
+EOF
+
+  # Add test toolchains using the constraints.
+  write_test_toolchain
+  cat >> BUILD <<EOF
+load('//toolchain:toolchain_test_toolchain.bzl', 'test_toolchain')
+
+# Define the toolchains.
+test_toolchain(
+    name = 'test_toolchain_impl_foo',
+    extra_str = 'foo',
+    visibility = ['//visibility:public'])
+
+test_toolchain(
+    name = 'test_toolchain_impl_bar',
+    extra_str = 'bar',
+    visibility = ['//visibility:public'])
+
+# Declare the toolchains.
+toolchain(
+    name = 'test_toolchain_foo',
+    toolchain_type = '//toolchain:test_toolchain',
+    exec_compatible_with = [],
+    target_compatible_with = [
+      # No constraint set, takes the default.
```

Comment:
What's the use case for defaults? Are they required for a constraint_setting or are they optional?

For example, if have `constraint_setting(name = "macos_version", default_constraint_value = ":osx10_11_6")` and we build for platform `"osx10_11_6"`, does that risk pulling in toolchains that don't even support OSX because they automatically take the default?

More generally, is there risk in a toolchain "indirectly" advertising support for a constraint the toolchain doesn't even know exists?

Reply:
Defaults are optional.

During toolchain resolution, only the constraint settings that are present in either the platform or the toolchain are considered. Therefore, if the platform does not declare a constraint value for "macos_version", and the toolchain doesn't declare a constraint value for "macos_version", the setting isn't considered. If either of them does use that setting, the other would be considered to have the default value, which the constraint author should have chosen as a safe value.

This feature is intended less for thing like "macos_version", and more for tunable constraints such as "compilation_mode" or "enable_race_detection", to be used as toolchain modes during resolution.

Reply:
Does "enable_race_detection" really need a default, or can its absence in a toolchain definition alone imply lack of support? Are there any precise motivating examples anywhere?

Sorry, I'm not trying to be difficult about this. There's a very real argument for defaults under circumstances where the alternative is having to, say, add a new mode that just one toolchain variant uses, then being forced to explicitly tag all other variants with the "old" mode setting. That can get unwieldy.

But I've also learned to be more cautious when using defaults. If not careful they provide behavior at a distance that's hard to track down and in a worst case toolchains, etc. being selected on properties the toolchain designers didn't necessarily even think of.

We've seen this tug and pull with existing compatible_with / restricted_to deployments.

Reply:
In addition from the uses rules_go wants, there are some uses that RBE wants as well. In both cases, the initial advice given was "use the absence of the value as the default", and that works okay, but in both cases users end up tagging several platform and toolchain definitions with comments like "constraint race_detection disabled by default", where it would be easier to just add the constraint value.

Also, having an explicit constraint value for the default makes it easier to switch the default later.